src/Microsoft.AspNetCore.Http.Connections/Internal/Transports/ServerSentEventsMessageFormatter.cs

```diff
@@ -27,37 +27,15 @@ public static class ServerSentEventsMessageFormatter
             // to the stream without modification. While it does mean that there will be smaller writes, should be fine for the most part
             // since we're using reasonably sized buffers.
 
-            if (payload.IsSingleSegment)
-            {
-                WriteMessage(payload.First, ms);
-            }
-            else
-            {
-                WriteMessage(payload.ToArray(), ms);
-            }
+            WriteMessage(payload, ms);
 
             ms.Position = 0;
```

Comment:
Why is a MemoryStream written to instead of writing directly to the response output stream? Is it because writing to the response output stream should use `WriteAsync` and didn't want the overhead of async while parsing?

Reply:
We can’t write partial messages to the JavaScript client. It can handle batches but not partial messages

Reply:
We could use our new more efficient memory stream. Orrr we could port pipelines to JavaScript 😬

Reply:
Yes also writing directly from the parsing loop would require it to be async.

Reply:
I actually think we should do a quick check for a new line. If there’s none then skip this whole thing

Reply:
Even if there was no new line, you would still need to write "data: " and "\r\n" with the content in one call.

Reply:
No that’s not a problem. It’s partial frames within the SSE frames. So you can write data first then the payload then \r\n. The browser (EventSource api) buffers until you get an entire frame.

Reply:
This is a problem for example:


data: { “target”: “Metho\r\n
data: d”, “args”: []\r\n
data: }\r\n

Reply:
Will that ever be a problem? JSON doesn't have raw \r\n in its content. It is always escaped when written inside a string.

As long as you aren't send pretty printed JSON (e.g. indented with new lines) then you'll never have that issue.

Reply:
Exactly why old SignalR worked because it was only json and we never had this logic. That’s why I want the fast path. The reason this logic exists here is because this transport layer doesn’t know we’re writing JSON.

Reply:
Still, I don't really like making this assumption in the transport if we can avoid it.

Reply:
Nah, we shouldn't I think we should do a PositionOf and if that returns null. Skip everything.